experimental/kernels/kernels_c.cpp

```diff
@@ -0,0 +1,525 @@
+#include "gpu.h"
+#include <array>
+#include <cstdio>
+#include <future>
+
+#include "kernels.h"
+#include "kernels_c.h"
+
+using namespace gpu; // createContext, createTensor, createKernel,
+                     // createShader, dispatchKernel, wait, toCPU
+                     // Tensor, Kernel, Context, Shape, kf32
+
+#define VOCAB_SIZE 50257
+
+// See https://github.com/google/dawn/blob/a8fbe981a86cb59536e2de423d2013a82d9b54a0/src/dawn/native/Limits.cpp
+#define LIMITS_BUFFER_SIZE_1GB { \
+    .limits = { \
+      .maxTextureDimension1D=8192, \
+      .maxTextureDimension2D=8192, \
+      .maxTextureDimension3D=2048, \
+      .maxTextureArrayLayers=256, \
+      .maxBindGroups=4, \
+      .maxBindGroupsPlusVertexBuffers=24, \
+      .maxBindingsPerBindGroup=1000, \
+      .maxDynamicUniformBuffersPerPipelineLayout=8, \
+      .maxDynamicStorageBuffersPerPipelineLayout=4, \
+      .maxSampledTexturesPerShaderStage=16, \
+      .maxSamplersPerShaderStage=16, \
+      .maxStorageBuffersPerShaderStage=8, \
+      .maxStorageTexturesPerShaderStage=4, \
+      .maxUniformBuffersPerShaderStage=12, \
+      .maxUniformBufferBindingSize=65536, \
+      .maxStorageBufferBindingSize=1073741824, \
+      .minUniformBufferOffsetAlignment=256, \
+      .minStorageBufferOffsetAlignment=256, \
+      .maxVertexBuffers=8, \
+      .maxBufferSize=0x80000000, \
+      .maxVertexAttributes=16, \
+      .maxVertexBufferArrayStride=2048, \
+      .maxInterStageShaderComponents=64, \
+      .maxInterStageShaderVariables=16, \
+      .maxColorAttachments=8, \
+      .maxColorAttachmentBytesPerSample=32, \
+      .maxComputeWorkgroupStorageSize=16384, \
+      .maxComputeInvocationsPerWorkgroup=256, \
+      .maxComputeWorkgroupSizeX=256, \
+      .maxComputeWorkgroupSizeY=256, \
+      .maxComputeWorkgroupSizeZ=64, \
+      .maxComputeWorkgroupsPerDimension=65535 \
+    }, \
+    .nextInChain = nullptr \
+  }
+
+void ENCODER_FORWARD_GPU(float* out,
```

Comment:
Any reason for all caps here? Could we use camelCase instead - the all caps makes these all look like macros.

(fwiw i personally prefer snake_case to camelCase but the webgpu API is all camelCase so we're going with it for consistency).

Also, instead of having the _GPU suffix, i would treat it as an implicit assumption that gpu implementations are the default, and if there's going to be a suffix, have it only on the cpu implementations.

Reply:
llm.c uses camelCase, so it uses camelCase.

`ENCODER_FORWARD_GPU` is used to replace it with `encoder_forward` in llm.c when used in a macro, and `encoder_forward_dummy` when not used.

`kernels_c.h` and `kernels_c.cpp` are intended to run the unit test(test_gpt2.c) in llm.c.

The kernels in llm.c are written in C. For example, the encoder kernel is `encoder_forward`. 
`kernels_c.h` provides the same kernel  interface as llm.c. The header is for C.
`kernels_c.h` is loaded into `train_gpt.c` in llm.c.

Efficiency and ease of use have been sacrificed in order to match C calls. 
For example, `kernels_c.h` and `kernels_c.cpp` do not have a Context as an argument, and create a Context unnecessarily.

Reply:
Since it would be difficult to test with train_gpt2.cu of CUDA version by replacing each function one by one, I used train_gpt2.c.
https://github.com/karpathy/llm.c/blob/master/train_gpt2.cu

Reply:
However, once the backward functions work, we can use the CUDA version.